SU2_CFD/include/gradients/computeGradientsLeastSquares.hpp

```diff
@@ -152,155 +146,190 @@ void computeGradientsLeastSquares(CSolver* solver,
+  }
 
-    su2double r11 = Rmatrix(iPoint,0,0);
-    su2double r12 = Rmatrix(iPoint,0,1);
-    su2double r22 = Rmatrix(iPoint,1,1);
-    su2double r13 = 0.0, r23 = 0.0, r23_a = 0.0, r23_b = 0.0, r33 = 0.0;
+}
+
+/*!
+ * \brief Solve the LS problem prepared above.
+ */
+template<class GradientType, class RMatrixType>
+FORCEINLINE void solveLeastSquares(size_t iPoint,
+                                   bool periodic,
+                                   size_t nDim,
+                                   size_t varBegin,
+                                   size_t varEnd,
+                                   const RMatrixType& Rmatrix,
+                                   GradientType& gradient) {
+  constexpr size_t MAXNDIM = 3;
 
+  /*--- Entries of upper triangular matrix R. ---*/
+
+  su2double r11 = Rmatrix(iPoint,0,0);
+  su2double r12 = Rmatrix(iPoint,0,1);
+  su2double r22 = Rmatrix(iPoint,1,1);
+  su2double r13 = 0.0, r23 = 0.0, r23_a = 0.0, r23_b = 0.0, r33 = 0.0;
+
+  if (periodic)
+  {
     AD::StartPreacc();
     AD::SetPreaccIn(r11);
     AD::SetPreaccIn(r12);
     AD::SetPreaccIn(r22);
+  }
 
-    if (r11 >= 0.0) r11 = sqrt(r11);
-    if (r11 >= 0.0) r12 /= r11; else r12 = 0.0;
-    su2double tmp = r22-r12*r12;
-    if (tmp >= 0.0) r22 = sqrt(tmp); else r22 = 0.0;
+  if (r11 >= 0.0) r11 = sqrt(r11);
+  if (r11 >= 0.0) r12 /= r11; else r12 = 0.0;
+  su2double tmp = r22-r12*r12;
+  if (tmp >= 0.0) r22 = sqrt(tmp); else r22 = 0.0;
 
-    if (nDim == 3) {
-      r13   = Rmatrix(iPoint,0,2);
-      r23_a = Rmatrix(iPoint,1,2);
-      r23_b = Rmatrix(iPoint,2,1);
-      r33   = Rmatrix(iPoint,2,2);
+  if (nDim == 3) {
+    r13   = Rmatrix(iPoint,0,2);
+    r23_a = Rmatrix(iPoint,1,2);
+    r23_b = Rmatrix(iPoint,2,1);
+    r33   = Rmatrix(iPoint,2,2);
 
+    if (periodic)
+    {
       AD::SetPreaccIn(r13);
       AD::SetPreaccIn(r23_a);
       AD::SetPreaccIn(r23_b);
       AD::SetPreaccIn(r33);
+    }
 
-      if (r11 >= 0.0) r13 /= r11; else r13 = 0.0;
-
-      if ((r22 >= 0.0) && (r11*r22 >= 0.0)) {
-        r23 = r23_a/r22 - r23_b*r12/(r11*r22);
-      } else {
-        r23 = 0.0;
-      }
+    if (r11 >= 0.0) r13 /= r11; else r13 = 0.0;
 
-      tmp = r33 - r23*r23 - r13*r13;
-      if (tmp >= 0.0) r33 = sqrt(tmp); else r33 = 0.0;
+    if ((r22 >= 0.0) && (r11*r22 >= 0.0)) {
+      r23 = r23_a/r22 - r23_b*r12/(r11*r22);
+    } else {
+      r23 = 0.0;
     }
 
-    /*--- Compute determinant ---*/
+    tmp = r33 - r23*r23 - r13*r13;
+    if (tmp >= 0.0) r33 = sqrt(tmp); else r33 = 0.0;
+  }
 
-    su2double detR2 = (r11*r22)*(r11*r22);
-    if (nDim == 3) detR2 *= r33*r33;
+  /*--- Compute determinant ---*/
 
-    /*--- Detect singular matrices ---*/
+  su2double detR2 = (r11*r22)*(r11*r22);
+  if (nDim == 3) detR2 *= r33*r33;
 
-    bool singular = false;
+  /*--- Detect singular matrix ---*/
 
-    if (detR2 <= EPS) {
-      detR2 = 1.0;
-      singular = true;
-    }
+  bool singular = (detR2 <= EPS);
```

Comment:
I see EPS is set to 1e-16. Perhaps we should base it on std::numeric_limits<T>::epsilon()